docs/design/RUC-on-type.md

```diff
@@ -0,0 +1,189 @@
+```C#
+[RequiresUnreferencedCode ("This method calls dangeours code")]
+public static void MethodWithDangerousCode () 
+{
+    // Dangerous code that generetares a trimming warning
+
+static void TestSuppressionOnMethod ()
+{
+    // IL2026: TestSuppressionOnMethod(): Using method 'MethodWithDangerousCode()' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. This method calls dangeours code.
+    MethodWithDangerousCode()
+}
+```
+
+Similarly RequiresUnreferencedCode on a type suppresses other warnings, with the difference that it will silence warnings from all methods in the type and attributes (except maybe static constructors, pending discussion)
+
+```C#
+[RequiresUnreferencedCode ("Message for --ClassWithRequiresUnreferencedCode--")]
+class ClassWithRequiresUnreferencedCode
+{
+    public static void MethodWithDangerousCode ()
+    { 
+        // Dangerous code that generetares a trimming warning
+    }
+}
+
+static void TestSuppressionOnType ()
+{
+    // IL2026: TestSuppressionOnMethod(): Using method 'ClassWithRequiresUnreferencedCode.MethodWithDangerousCode()' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Message for --ClassWithRequiresUnreferencedCode--.
+    ClassWithRequiresUnreferencedCode.MethodWithDangerousCode()
+}
+```
+
+## The behavior of Nested Classes
+We decided to exclude the auto propagation of the attribute to nested classes, this means that a nested class is not considered to have RequiresUnreferencedCode if inside a type marked with RequiresUnreferencedCode
+
+```C#
+[RequiresUnreferencedCode ("Message for --ClassWithRequiresUnreferencedCode--")]
+class ClassWithRequiresUnreferencedCode
+{
+    public static void StaticMethod () { }
+    class NestedClass {
+        public static void StaticMethod () { }
+    }
+}
+
+static void TestRequiresUnreferencedCodeInClass ()
+{
+    // IL2026: TestRequiresUnreferencedCodeInClass(): Using method 'ClassWithRequiresUnreferencedCode.StaticMethod()' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Message for --ClassWithRequiresUnreferencedCode--.
+    ClassWithRequiresUnreferencedCode.StaticMethod ();
+
+    ClassWithRequiresUnreferencedCode.NestedClass.StaticMethod ();
+}
+
+```
+In the previous example calling ClassWithRequiresUnreferencedCode.StaticMethod () is considered dangerous since the declaring type of the static method is annotated with RequiresUnreferencedCode. For ClassWithRequiresUnreferencedCode.NestedClass.StaticMethod () the declared type NestedClass is not considered dangerous since it doesn't contain the RequiresUnreferencedCode on it nor the method contains the attribute.
+
+If we want ClassWithRequiresUnreferencedCode.NestedClass.StaticMethod () to generate the warning we would need to annotate either the NestedClass type or the ClassWithRequiresUnreferencedCode.NestedClass.StaticMethod static method.
```

Comment:
I think much of this text is more of a guide than a spec. There's nothing wrong with guides, but specs can be much shorter and just list the exact behavior that the implementers need to know.